src/doc/grammar.dox

```diff
@@ -0,0 +1,384 @@
+ to skip creating G.fst and just create LG.fst manually; this won't be hard to
+ do once you know its expected structure.
+
+  \section grammar_special_lg  Special symbols in LG.fst
+
+ Before we describe what L.fst is expected to do with the special symbols,
+ we will state what we expect LG.fst to contain after composition.  All the
+ special symbols are on the ilabels.
+
+ Let us define the set of <e>"left-context phones"</e> as the set of phones that can
+ end a word, plus the optional silence, plus the special symbol `\#nonterm_bos`.
+ This is the set of phones that can possibly appear as the left-context when we
+ are beginning a word, plus `\#nonterm_bos` as a stand-in for the beginning-of-sequence
+ context where no previous phone was seen.  We will italicize the phrase
+ <e>left-context phones</e> when we use it, to emphasize that it has a special meaning.
+
+ For non-top-level graphs only:
+
+  - All paths in FST must begin with `\#nonterm_begin` followed by each possible
+    <e>left-context phone</e>, i.e. parallel arcs enumerating all possible phonetic
+    left-contexts that could precede this nonterminal.
+
+    In non-word-position-dependent systems we can just let this set be all phones;
+    in word-position-dependent system it can be all phones except word-internal
+    and word-begin phones, i.e.  all phones except those that look like `P_B`
+    and `P_I`.  If the set of possible left contexts is known to be smaller, it may
+    be more efficient to make this a smaller set.  In addition to real phones, if
+    it is possible for this nonterminal to be the first phone in an utterance it is
+    necessary for the set of phones mentioned above to include `\#nonterm_bos`.
+
+  - ilabel sequences must end with #nonterm_end.
+
+ Whenever a nonterminal is invoked, whether from a top-level or non-top-level
+ graph, the ilabels in LG.fst will be, for example, `\#nonterm:foo` followed by
+ in parallel, all possible <e>left-context phones</e>.
+
+
+  \section grammar_special_l  Special symbols in L.fst
+
+ This section explains what sequences involving special symbols in L.fst we need to
+ add, in order to compile a LG.fst with the desired properties from G.fst.  Let the
+ loop-state of L.fst be the state in L.fst with very high out-degree, from which
+ all the words leave (and return).
+
+ The lexicon needs to include, in addition to the normal things:
+
+ - A sequence starting and ending at the loop-state, with olabel `\#nonterm_begin`
+   and ilabels consisting of, `\#nonterm_begin` followed by all possible left-context
+   phones (and `\#nonterm_bos`) in parallel.
+ - A self-loop arc at the loop-state, with olabel `\#nonterm_end`
+   and ilabel `\#nonterm_end`.
+ - For each user-defined nonterminal (e.g. `\#nonterm:foo`) and for
+   `\#nonterm_begin`, the following:
+    - An arc from the loop-state to a newly created state, with ilabel
+      and olabel equal to the symbol in question (say `\#nonterm:foo`).
```

Comment:
@ngoel17 
This is just a mechanism to pass those symbols through the lexicon, so if they appear on the ilabels of the grammar, they will also appear on the ilabels of LG.fst; and to add all possible phonetic contexts where appropriate.
I'm not sure what scenario you want to support in combination with this?  An example might help

Reply:
Example is : input:  one hundred and seven
                   output:  107
   It can be achieved by having a G.fst where ilable and olabel are not the same.